_posts/2024-12-03-safety-of-cstr.md

```diff
@@ -0,0 +1,539 @@
+
+#### Verification Checks
+
+After method invocation, we had two checks:
+
+1. **Correctness Checks**: Ensure that the function returns `Ok` instead of an error when an valid input is given. An `Ok` result simply indicates that a `CStr` instance was successfully created; it does not inherently guarantee the safety of the `CStr`.
+2. **Safety Checks**: Verify that the resulting `CStr` satisfies the safety invariant by calling `is_safe()`.
+
+#### Performance Improvement
+
+You may have noticed the use of `#[kani::unwind(32)]` in the harness. This was necessary since we were working with a `MAX_SIZE` of `32` bytes. The unwinding bound must account for:
+
+- The main loop that processes the bytes, and
+- Any additional iterations for safety checks, such as searching for null terminators.
+
+In some cases, we need to unwind one extra iteration, such as with `#[kani::unwind(33)]`, to verify the presence of a null terminator (at position 32). This is especially crucial for functions that:
+
+- Scan for null bytes (e.g. `strlen`)
+- Verify string boundaries
+- Check array indices up to and including the null terminator
+
+The unwinding bound must be greater and equal to `MAX_SIZE + 1` to ensure complete verification of all possible execution paths, including edge cases involving the null terminator.
+
+You can find more about [loop unwinding](https://model-checking.github.io/kani/tutorial-loop-unwinding.html) in the Kani tutorial.
+
+### Interlude: Helper Function `arbitray_cstr`
+
+In many cases, we needed to verify `CStr` methods that operate directly on a `CStr` object itself. To achieve this, we would like to generate an arbitrary `CStr` instance for verification, similar to how we generated input slices for `from_bytes_with_nul`.
+
+The following code shows the implementation of `arbitray_cstr`:
+
+```rust
+fn arbitrary_cstr(slice: &[u8]) -> &CStr {
+	// At a minimum, the slice has a null terminator to form a valid CStr.
+	kani::assume(slice.len() > 0 && slice[slice.len() - 1] == 0);          // (1)
+	let result = CStr::from_bytes_until_nul(&slice);                       // (2)
+	// Given the assumption, from_bytes_until_nul should never fail
+        assert!(result.is_ok());                                               // (3)
+	let c_str = result.unwrap();
+	assert!(c_str.is_safe());                                              // (4)
+	c_str
+}
+```
+
+`arbitray_cstr` consists of four key steps:
+
+1. **Assumption**: Assumes that the input slice is non-empty and ends with a null terminator. This is a small optimization which guarantees that the input slice contains at least a null terminator.
```

Comment:
> I'm also curious why you didn't implement Kani's Arbitrary [trait](https://model-checking.github.io/kani/crates/doc/kani/trait.Arbitrary.html) instead.

I tried to implement `Arbitrary` for `CStr` before, but the compilation failed because `CStr` does not implement the `Sized` trait, so its size is unknown during compilation time.